Classes/Service/NodePolicyService.php

```diff
@@ -114,14 +115,68 @@ public function getDisallowedNodeTypes(NodeInterface $node): array
             return $disallowedNodeTypes;
         }
 
+        // determine the set of configured node supertypes
+        $nodeNodeType = $node->getNodeType();
+
+        $superTypes = [];
+        $generateSuperTypes = function (array $nodeTypes, &$superTypes) use (&$generateSuperTypes) {
+            foreach ($nodeTypes as $nodeType) {
+                $superTypes[$nodeType->getName()] = $nodeType;
+                $generateSuperTypes($nodeType->getDeclaredSuperTypes(), $superTypes);
+            }
+        };
+        $generateSuperTypes($nodeNodeType->getDeclaredSuperTypes(), $superTypes);
+
+        $parentNodeType = null;
+        try {
+            $parentNode = $node->findParentNode();
+            $parentNodeType = $parentNode->getNodeType();
+        } catch (NodeException $exception) {
+            // no parentNode found
+        }
+
+        // check if the node is auto-created
+        $isAutCreated   = false;
+        if ($parentNodeType && array_key_exists((string)$node->getNodeName(), $parentNodeType->getAutoCreatedChildNodes())) {
+            $isAutCreated = true;
+        }
+
+        // filter the set of configured nodeTypes
+        $nodeName = (string)$node->getNodeName();
+
+        $constraintAndSuperTypeFilter = function ($nodeType) use ($nodeName, $nodeNodeType, $superTypes, $isAutCreated, $parentNodeType) {
+            // check if the nodeType is mentioned in the constraints
+            if ($isAutCreated) {
+                if ($parentNodeType && $parentNodeType->allowsGrandchildNodeType($nodeName, $nodeType)) {
+                    return true;
+                }
+            } elseif ($nodeNodeType->allowsChildNodeType($nodeType)) {
+                return true;
+            }
+            // check if the nodeType is a supertype
+            elseif (isset($superTypes[$nodeType->getName()])) {
+                return true;
+            }
+            // check if the nodeType is the same as that of the current $node's nodeType
+            elseif ($nodeType->getName() == $nodeNodeType->getName()) {
+                return true;
+            }
+
+            // ignore other nodeType
+            return false;
+        };
+
+        $nodeTypes = array_filter($this->nodeTypeManager->getNodeTypes(), $constraintAndSuperTypeFilter);
```

Comment:
For my taste I'd call it smth like `filteredNodeTypes`

Reply:
Yes, you are right. I've changed the variable name according to your suggestion.